aten/src/ATen/native/quantized/cudnn/Linear.cpp

```diff
@@ -0,0 +1,345 @@
+#include <ATen/TensorUtils.h>
+#include <c10/core/ScalarType.h>
+#include <cudnn_frontend.h>
+#include <torch/library.h>
+
+#include <iostream>
+#include <unordered_map>
+
+// TODO: there is a table from input dtype and weight dtype to operator dtype,
+// we can derive the operator dtype based on input dtype
+cudnn_frontend::MatMulDesc_v8 getLinearDescriptor(cudnnDataType_t dataType) {
+  return cudnn_frontend::MatMulDescBuilder()
+    .setMathPrecision(dataType)
+    .build();
+}
+
+struct CacheKey {
+  uint8_t input_alignment;
+  uint8_t weight_alignment;
+  uint8_t output_alignment;
+  // default to -1 when no bias
+  int8_t bias_alignment;
+};
+
+// FIXME: make this thread-safe by reusing the benchmark cache in Conv_v7.cpp
+namespace {
+std::unordered_map<CacheKey, cudnn_frontend::ManagedOpaqueDescriptor, at::native::ParamsHash<CacheKey>, at::native::ParamsEqual<CacheKey>> execution_plan_cache;
+}
+// TODO: we can use cudnn_frontend::ExecutionPlanCache when it supports caching
+// multiple operators
+// reference: https://github.com/NVIDIA/cudnn-frontend/blob/main/samples/conv_sample.cpp#L293
+//static cudnn_frontend::ExecutionPlanCache plan_cache("sample_cache");
+
+// currently we only support int8 symmetric (zero_point = 0 for inputs and output) quantized linear op
+// We implement relu(act_int8 * transpose(w_int8) + [bias_fp32/(act_scale * w_scale] ) * ( act_scale * w_scale / out_scale )
+// which requires 5 cudnn ops (1 matmul, 2 multiplication, 1 add, and 1 relu ops)
+// matmul op: linear_op
+// Multiplication ops: rhs_mult_op, requant_op
+// Addition op: add_op
+// Relu op: relu_op
+template <bool kReluFused>
+void PackedLinearWeightCudnn::apply_impl_helper(const at::Tensor& quantized_output, const at::Tensor& input, double output_scale) {
+  if (quantized_output.numel() == 0) {
+    return;
+  }
+  at::Tensor linear_output = at::empty(quantized_output.sizes(), at::device(at::kCUDA).dtype(at::kFloat));
+  // TODO: combine empty & fill_ using full_like or full
+  at::Tensor requantize_multiplier_tensor = at::empty(quantized_output.sizes(), at::device(at::kCUDA).dtype(at::kFloat));
+  auto act_scale = input.q_scale();
+  auto weight_scale = orig_weight.q_scale();
+  auto requantize_multiplier = act_scale * weight_scale / output_scale;
+  requantize_multiplier_tensor.fill_(requantize_multiplier);
+  c10::optional<at::Tensor> bias_multiplier_tensor;
+  c10::optional<at::Tensor> broadcasted_bias;
+  if (bias_.has_value()) {
+    // the input bias is a 1-D tensor whose size is the same as the size of the second dimension of quantized_output.
+    // we need to add trailing dimensions in order to properly broadcast bias, otherwise broadcast_to will fail.
+    // the number of trailling dimensions is quantized_output.dim() - 2, so the new size of the broadcast_bias
+    // becomes quantized_output.dim() - 2 + 1. nothing needs to be done for the leading dimensions
+    std::vector<int64_t> new_size(quantized_output.dim() - 1, 1);
```

Comment:
1. is the call `ndim`?
2. why this is `quantized_output.dim() - 1`? shouldn't it match the dimension of quantized_output?